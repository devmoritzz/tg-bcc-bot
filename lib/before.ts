import * as Path from 'path'
import * as dotenv from 'dotenv-safe'

process.on('unhandledRejection', err => {
    console.error(err)
    process.exit(1)
})

dotenv.config({
    path: Path.resolve(__dirname, '../../dotenv'),
    example: Path.resolve(__dirname, '../../dotenv.example'),
})
