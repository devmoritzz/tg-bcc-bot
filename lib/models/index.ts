import * as Knex from 'knex'
import cfg from './cfg'
import lazy from '../utils/lazy'

const conn = lazy(() => Knex(cfg))

export default {
    async init() {},
    async destroy() {
        await conn().destroy()
    },

    async addTags(chat: number, tags: string[]): Promise<void> {
        if (tags.length === 0) return
        const vals = tags.map(tag => ({ chat, tag }))
        const insert = conn()
            .insert(vals)
            .toString()
        const query = `${insert} ON CONFLICT DO NOTHING`
        await conn().raw(query)
    },

    async getTags(chat: number): Promise<string[]> {
        const r = await conn()
            .select('tag')
            .from('tg_chat_tag')
            .where({ chat })
        return r
    },
}

export interface ChatTag {
    id: number
    chat: number
    tag: string
}
