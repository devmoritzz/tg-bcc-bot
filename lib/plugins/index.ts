import logger from './logger'
import rollbar from './rollbar'

const register = async server => {
    await server.register(logger)
    await server.register(rollbar)
}

export default register
