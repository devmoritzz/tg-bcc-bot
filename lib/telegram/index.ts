export const reply = (_payload:object) => {}
