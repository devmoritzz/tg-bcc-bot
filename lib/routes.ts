import * as TG from './controllers/telegram'

const routes = [
    { path: '/tg/webhook', method: 'post', options: TG.webhook },
].filter(Boolean)

export default routes
