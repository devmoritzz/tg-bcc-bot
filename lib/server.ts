import * as Path from 'path'
import * as Hapi from '@hapi/hapi'
import * as dotenv from 'dotenv-safe'
import { routes } from './routes'
import plugins from './plugins'
import model from './models'
import { setWebhook } from './telegram/api'

process.on('unhandledRejection', err => {
    console.error(err)
    process.exit(1)
})

const common = async () => {
    dotenv.config({
        path: Path.resolve(__dirname, '../../dotenv'),
        example: Path.resolve(__dirname, '../../dotenv.example'),
    })

    await model.init()

    const server = Hapi.server({
        port: Number(process.env.PORT || 8000),
        host: '0.0.0.0',
        routes: {
            payload: { allow: 'application/json' },
        },
    })
    await plugins(server)
    server.route(routes)

    server.events.on('stop', async () => {
        await model.destroy()
    })

    return server
}

export const init = async () => {
    const server = await common()
    server.initialize()
    return server
}

export const start = async () => {
    const server = await common()
    await server.start()

    const url = `https://${process.env.TG_WEBHOOK_DOMAIN}/tg/webhook`
    await setWebhook({ url })

    return server
}
