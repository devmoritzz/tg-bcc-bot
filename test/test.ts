import { init } from '../lib/server'

let server: any
beforeAll(async () => {
    server = await init()
})
afterAll(async () => {
    if (server) {
        await server.stop()
    }
})

describe('webhook', () => {
    const url = process.env.TG_WEBHOOK_PATH!
    // https://core.telegram.org/bots/webhooks#testing-your-bot-with-updates
    test('message with text', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    type: 'private',
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                text: '/start',
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('forwarded message', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    type: 'private',
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                forward_from: {
                    last_name: 'Forward Lastname',
                    id: 222222,
                    first_name: 'Forward Firstname',
                },
                forward_date: 1441645550,
                text: '/start',
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('forwarded channel message', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                forward_from: {
                    id: -10000000000,
                    type: 'channel',
                    title: 'Test channel',
                },
                forward_date: 1441645550,
                text: '/start',
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('message with a reply', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                text: '/start',
                reply_to_message: {
                    date: 1441645000,
                    chat: {
                        last_name: 'Reply Lastname',
                        type: 'private',
                        id: 1111112,
                        first_name: 'Reply Firstname',
                        username: 'Testusername',
                    },
                    message_id: 1334,
                    text: 'Original',
                },
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('edited message', async () => {
        const payload = {
            update_id: 10000,
            edited_message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                text: 'Edited text',
                edit_date: 1441646600,
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('message with entities', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                text: 'Bold and italics',
                entities: [
                    {
                        type: 'italic',
                        offset: 9,
                        length: 7,
                    },
                    {
                        type: 'bold',
                        offset: 0,
                        length: 4,
                    },
                ],
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('message with audio', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                audio: {
                    file_id: 'AwADBAADbXXXXXXXXXXXGBdhD2l6_XX',
                    duration: 243,
                    mime_type: 'audio/mpeg',
                    file_size: 3897500,
                    title: 'Test music file',
                },
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('voice message', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                voice: {
                    file_id: 'AwADBAADbXXXXXXXXXXXGBdhD2l6_XX',
                    duration: 5,
                    mime_type: 'audio/ogg',
                    file_size: 23000,
                },
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('message with a document', async () => {
        const payload = {
            update_id: 10000,
            message: {
                date: 1441645532,
                chat: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                message_id: 1365,
                from: {
                    last_name: 'Test Lastname',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                document: {
                    file_id: 'AwADBAADbXXXXXXXXXXXGBdhD2l6_XX',
                    file_name: 'Testfile.pdf',
                    mime_type: 'application/pdf',
                    file_size: 536392,
                },
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('inline query', async () => {
        const payload = {
            update_id: 10000,
            inline_query: {
                id: 134567890097,
                from: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                query: 'inline query',
                offset: '',
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('chosen inline query', async () => {
        const payload = {
            update_id: 10000,
            chosen_inline_result: {
                result_id: '12',
                from: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                query: 'inline query',
                inline_message_id: '1234csdbsk4839',
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('callback query', async () => {
        const payload = {
            update_id: 10000,
            callback_query: {
                id: '4382bfdwdsb323b2d9',
                from: {
                    last_name: 'Test Lastname',
                    type: 'private',
                    id: 1111111,
                    first_name: 'Test Firstname',
                    username: 'Testusername',
                },
                data: 'Data from button callback',
                inline_message_id: '1234csdbsk4839',
            },
        }

        const resp = await server.inject({
            method: 'post',
            url,
            headers: {
                'Content-Type': 'application/json',
                'Cache-Control': 'no-cache',
            },
            payload,
        })
        expect(resp.result).toMatchSnapshot()
    })
})
