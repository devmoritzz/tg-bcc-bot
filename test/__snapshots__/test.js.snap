// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhook callback query 1`] = `"ok"`;

exports[`webhook chosen inline query 1`] = `"ok"`;

exports[`webhook edited message 1`] = `"ok"`;

exports[`webhook forwarded channel message 1`] = `"ok"`;

exports[`webhook forwarded message 1`] = `"ok"`;

exports[`webhook inline query 1`] = `"ok"`;

exports[`webhook message with a document 1`] = `"ok"`;

exports[`webhook message with a reply 1`] = `"ok"`;

exports[`webhook message with audio 1`] = `"ok"`;

exports[`webhook message with entities 1`] = `"ok"`;

exports[`webhook message with text 1`] = `"ok"`;

exports[`webhook voice message 1`] = `"ok"`;
